xtable-core/src/test/java/org/apache/xtable/delta/ITDeltaDeleteVectorConvert.java

```diff
@@ -0,0 +1,155 @@
+    hadoopConf.set("fs.defaultFS", "file:///");
+
+    conversionSourceProvider = new DeltaConversionSourceProvider();
+    conversionSourceProvider.init(hadoopConf);
+  }
+
+  @Test
+  public void testInsertsUpsertsAndDeletes() {
+    String tableName = GenericTable.getTableName();
+    TestSparkDeltaTable testSparkDeltaTable =
+        new TestSparkDeltaTable(tableName, tempDir, sparkSession, null, false);
+
+    List<List<String>> allActiveFiles = new ArrayList<>();
+    List<TableChange> allTableChanges = new ArrayList<>();
+    List<Row> rows = testSparkDeltaTable.insertRows(50);
+    Long timestamp1 = testSparkDeltaTable.getLastCommitTimestamp();
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+
+    // enable deletion vectors for the test table
+    testSparkDeltaTable
+        .getSparkSession()
+        .sql(
+            "ALTER TABLE "
+                + tableName
+                + " SET TBLPROPERTIES ('delta.enableDeletionVectors' = true)");
+
+    List<Row> rows1 = testSparkDeltaTable.insertRows(50);
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(100L, testSparkDeltaTable.getNumRows());
+
+    // upsert does not create delete vectors
+    testSparkDeltaTable.upsertRows(rows.subList(0, 20));
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(100L, testSparkDeltaTable.getNumRows());
+
+    testSparkDeltaTable.insertRows(50);
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(150L, testSparkDeltaTable.getNumRows());
+
+    // delete a few rows with gaps in ids
+    List<Row> rowsToDelete =
+        rows1.subList(0, 10).stream()
+            .filter(row -> (row.get(0).hashCode() % 2) == 0)
+            .collect(Collectors.toList());
+    rowsToDelete.addAll(rows.subList(35, 45));
+    testSparkDeltaTable.deleteRows(rowsToDelete);
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(135L, testSparkDeltaTable.getNumRows());
+
+    testSparkDeltaTable.insertRows(50);
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(185L, testSparkDeltaTable.getNumRows());
+
+    // delete a few rows from a file which already has a deletion vector, this should generate a
+    // merged deletion vector file. Some rows were already deleted in the previous delete step.
+    // This deletion step intentionally deletes the same rows again to test the merge.
+    rowsToDelete = rows1.subList(5, 15);
+    testSparkDeltaTable.deleteRows(rowsToDelete);
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(178L, testSparkDeltaTable.getNumRows());
+
+    testSparkDeltaTable.insertRows(50);
+    allActiveFiles.add(testSparkDeltaTable.getAllActiveFiles());
+    assertEquals(228L, testSparkDeltaTable.getNumRows());
+
+    // TODO conversion fails if delete vectors are enabled, this is because of missing handlers for
```

Comment:
Can you explain what this TODO comment is about ? I'm new to delta deletion vectors.

Reply:
Currently, XTable doesn't identify and convert deletion vectors. There are two different representations of delete vectors in Delta and Iceberg today, and we can expect new representation to be added in the coming months. XTable will need to understand and convert all of these. Right now, XTable conversion isn't accurate because it lacks deletion vector support. I'm collaborating with new community members who are interested in adding this support. This change provides some utilities to help them get started. As we make progress, I can work on a document to explain the current formats and how XTable handles them.